kbkdf/src/lib.rs

```diff
@@ -0,0 +1,511 @@
+/// It takes multiple arguments:
+///  - Prf - the Pseudorandom Function to derive keys from
+///  - K - the expected output length of the newly derived key
+///  - R - An integer (1 <= r <= 32) that indicates the length of the binary encoding of the counter i
+///        as an integer in the interval [1, 2r − 1].
+pub trait Kbkdf<Prf, K, R: sealed::R>
+where
+    Prf: Mac + KeyInit,
+    K: KeySizeUser,
+    K::KeySize: ArrayLength<u8> + Mul<U8>,
+    <K::KeySize as Mul<U8>>::Output: Unsigned,
+    Prf::OutputSize: ArrayLength<u8> + Mul<U8>,
+    <Prf::OutputSize as Mul<U8>>::Output: Unsigned,
+{
+    /// Derives `key` from `kin` and other parameters.
+    fn derive(
+        &self,
+        //kin: &GenericArray<u8, Prf::KeySize>,
+        kin: &[u8],
+        use_l: bool,
+        use_separator: bool,
+        label: &[u8],
+        context: &[u8],
+    ) -> Result<GenericArray<u8, K::KeySize>, Error> {
+        // n - An integer whose value is the number of iterations of the PRF needed to generate L
+        // bits of keying material
+        let n: u32 = Wrapping(<KbkdfCore<K::KeySize, Prf::OutputSize> as KbkdfUser>::L::U32)
+            .div_ceil(Wrapping(
+                <KbkdfCore<K::KeySize, Prf::OutputSize> as KbkdfUser>::H::U32,
+            ))
+            .0;
+
+        if n as usize > 2usize.pow(R::U32) - 1 {
+            return Err(Error::InvalidRequestSize);
+        }
+
+        let mut output = GenericArray::<u8, K::KeySize>::default();
+        let mut builder = output.as_mut_slice();
+
+        let mut ki = None;
+        self.input_iv(&mut ki);
+        let mut a = {
+            let mut h = <Prf as Mac>::new_from_slice(kin).unwrap();
+            h.update(label);
+            if use_separator {
+                h.update(&[0]);
+            }
+            h.update(context);
+            h.finalize().into_bytes()
+        };
+
+        for counter in 1..=n {
+            if counter > 1 {
+                a = {
+                    let mut h = <Prf as Mac>::new_from_slice(kin).unwrap();
+                    h.update(a.as_slice());
+                    h.finalize().into_bytes()
+                };
+            }
+
+            let mut h = <Prf as Mac>::new_from_slice(kin).unwrap();
+
+            if Self::FEEDBACK_KI {
+                if let Some(ki) = ki {
+                    h.update(ki.as_slice());
+                }
+            }
+
+            if Self::DOUBLE_PIPELINE {
+                h.update(a.as_slice());
+            } else {
+                // counter encoded as big endian u32
+                // Type parameter R encodes how large the value is to be (either U8, U16, U24, or U32)
+                //
+                // counter = 1u32 ([0, 0, 0, 1])
+                //                     \-------/
+                //                      R = u24
+                h.update(&counter.to_be_bytes()[(4 - R::USIZE / 8)..]);
+            }
+
+            // Fixed input data
```

Comment:
According to the specification, the counter is hashed before the fixed data. But NIST test vectors have samples for `AFTER_FIXED`, `MIDDLE_FIXED`, and `BEFORE_ITER` counter locations. I'm wondering why is that.

I checked KBKDF implementation in other languages. For example, a `CounterLocation` is implemented in [Python's implementation](https://cryptography.io/en/latest/hazmat/primitives/key-derivation-functions/#cryptography.hazmat.primitives.kdf.kbkdf.CounterLocation). Should I implement it too or leave it as it is?